src/w3d/renderer/meshmdl.cpp

```diff
@@ -681,9 +681,7 @@ TextureClass *Load_Texture(ChunkLoadClass &cload)
 W3DErrorType MeshModelClass::Read_Textures(ChunkLoadClass &cload, MeshLoadContextClass *context)
 {
     for (TextureClass *newtex = Load_Texture(cload); newtex != nullptr; newtex = Load_Texture(cload)) {
-
         context->Add_Texture(newtex);
-        newtex->Release_Ref();
```

Comment:
```
==21136==ERROR: AddressSanitizer: heap-use-after-free on address 0x17c677c4 at pc 0x0083d709 bp 0x0133cf9c sp 0x0133cf90
READ of size 4 at 0x17c677c4 thread T0
    #0 0x83d708 in W3DAssetManager::Release_All_Textures D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:404
    #1 0x83a959 in W3DAssetManager::Free_Assets D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:197
    #2 0x83142e in CW3DViewDoc::OnNewDocument D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\w3dviewdoc.cpp:110
    #3 0x593a575d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1023575d)
    #4 0x593a52b2  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102352b2)
    #5 0x593a4c6d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10234c6d)
    #6 0x593529b9  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101e29b9)
    #7 0x59407a7a  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10297a7a)
    #8 0x81a5ec in CMainFrame::OnCmdMsg D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:195
    #9 0x593fa978  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028a978)
    #10 0x59406d2c  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10296d2c)
    #11 0x81a9e2 in CMainFrame::OnCommand D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:216
    #12 0x593f9f48  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289f48)
    #13 0x593f9ecc  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289ecc)
    #14 0x8213b5 in CMainFrame::WindowProc D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:229
    #15 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #16 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #17 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #18 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #19 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #20 0x76368a58 in GetParent+0x928 (C:\WINDOWS\System32\USER32.dll+0x69e38a58)
    #21 0x76365e3c in SendMessageW+0x4c (C:\WINDOWS\System32\USER32.dll+0x69e35e3c)
    #22 0x6f324d8b in CreateToolbar+0x4beb (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bf94d8b)
    #23 0x6f325cc2 in CreateToolbar+0x5b22 (C:\WINDOWS\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.4355_none_c0dc01d438beab35\COMCTL32.dll+0x5bf95cc2)
    #24 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #25 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #26 0x763672f5 in GetSystemMetricsForDpi+0xc45 (C:\WINDOWS\System32\USER32.dll+0x69e372f5)
    #27 0x763543ca in CallWindowProcA+0x1a (C:\WINDOWS\System32\USER32.dll+0x69e243ca)
    #28 0x593f8e23  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288e23)
    #29 0x593f9eed  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289eed)
    #30 0x59358fa6  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101e8fa6)
    #31 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #32 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #33 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #34 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #35 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)
    #36 0x76366470 in DispatchMessageW+0x4f0 (C:\WINDOWS\System32\USER32.dll+0x69e36470)
    #37 0x7636a3e7 in IsDialogMessageW+0x107 (C:\WINDOWS\System32\USER32.dll+0x69e3a3e7)
    #38 0x7635b0ed in IsDialogMessage+0x4d (C:\WINDOWS\System32\USER32.dll+0x69e2b0ed)
    #39 0x59410b4d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102a0b4d)
    #40 0x593fd426  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028d426)
    #41 0x59358f42  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101e8f42)
    #42 0x593fb4ff  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028b4ff)
    #43 0x593e51e9  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102751e9)
    #44 0x593e5b2a  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10275b2a)
    #45 0x593e5254  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10275254)
    #46 0x593e50ea  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102750ea)
    #47 0x5940e243  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1029e243)
    #48 0x9a543a in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #49 0x7696fcc8 in BaseThreadInitThunk+0x18 (C:\WINDOWS\System32\KERNEL32.DLL+0x6b81fcc8)
    #50 0x773f80cd in RtlGetAppContainerNamedObjectPath+0x11d (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4b2e80cd)
    #51 0x773f809d in RtlGetAppContainerNamedObjectPath+0xed (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4b2e809d)

0x17c677c4 is located 4 bytes inside of 112-byte region [0x17c677c0,0x17c67830)
freed by thread T0 here:
    #0 0x5965947d in __asan_wrap_RtlValidateHeap+0x196d (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\clang_rt.asan_dynamic-i386.dll+0x1003947d)
    #1 0x5966a41c in _asan_wrap__except_handler4+0x47ec (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\clang_rt.asan_dynamic-i386.dll+0x1004a41c)
    #2 0x9a4683 in operator delete D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win_delete_scalar_size_thunk.cpp:44
    #3 0x8b5698 in TextureClass::`scalar deleting destructor'+0x48 (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\w3dview.exe+0x535698)
    #4 0x82e396 in RefCountClass::Delete_This D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:53
    #5 0x7e5340 in RefCountClass::Release_Ref D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:49
    #6 0x7f41e7 in Ref_Ptr_Release<TextureClass> D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:81
    #7 0x8fc75d in MeshMatDescClass::Reset D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmatdesc.cpp:405
    #8 0x8a9431 in MeshModelClass::Reset D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmdl.cpp:108
    #9 0x8a0fd6 in MeshModelClass::~MeshModelClass D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmdl.cpp:67
    #10 0x8a215a in MeshModelClass::`scalar deleting destructor'+0xa (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\w3dview.exe+0x52215a)
    #11 0x82e396 in RefCountClass::Delete_This D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:53
    #12 0x7e5340 in RefCountClass::Release_Ref D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:49
    #13 0x87e218 in Ref_Ptr_Release<MeshModelClass> D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:81
    #14 0x8818de in MeshClass::Free D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\mesh.cpp:106
    #15 0x87ff1a in MeshClass::`scalar deleting destructor'+0xa (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\w3dview.exe+0x4fff1a)
    #16 0x82e396 in RefCountClass::Delete_This D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:53
    #17 0x7e5340 in RefCountClass::Release_Ref D:\Projects\TheAssemblyArmada\Thyme\src\w3d\lib\refcount.h:49
    #18 0x87ffff in PrimitivePrototypeClass::`scalar deleting destructor'+0xbf (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\w3dview.exe+0x4fffff)
    #19 0x8818c7 in PrimitivePrototypeClass::Delete_Self D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\mesh.h:148
    #20 0x83a8d3 in W3DAssetManager::Free_Assets D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:190
    #21 0x83142e in CW3DViewDoc::OnNewDocument D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\w3dviewdoc.cpp:110
    #22 0x593a575d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1023575d)
    #23 0x593a52b2  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102352b2)
    #24 0x593a4c6d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10234c6d)
    #25 0x593529b9  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101e29b9)
    #26 0x59407a7a  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10297a7a)
    #27 0x81a5ec in CMainFrame::OnCmdMsg D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:195

previously allocated by thread T0 here:
    #0 0x5965957d in __asan_wrap_RtlValidateHeap+0x1a6d (D:\Projects\TheAssemblyArmada\Thyme\.solution\vs2022\RelWithDebInfo\clang_rt.asan_dynamic-i386.dll+0x1003957d)
    #1 0x591d9035  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10069035)
    #2 0x83baf8 in W3DAssetManager::Get_Texture D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:377
    #3 0x8a5259 in Load_Texture D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmdl.cpp:622
    #4 0x8a897e in MeshModelClass::Read_Textures D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmdl.cpp:683
    #5 0x8a6c10 in MeshModelClass::Read_Chunks D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmdl.cpp:515
    #6 0x8a5a05 in MeshModelClass::Load_W3D D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\meshmdl.cpp:453
    #7 0x8829e8 in MeshClass::Load_W3D D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\mesh.cpp:525
    #8 0x882d82 in MeshLoaderClass::Load_W3D D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\mesh.cpp:793
    #9 0x83cb82 in W3DAssetManager::Load_Prototype D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:676
    #10 0x83c897 in W3DAssetManager::Load_3D_Assets D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:174
    #11 0x83caa3 in W3DAssetManager::Load_3D_Assets D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:145
    #12 0x830e67 in CW3DViewDoc::LoadFile D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\w3dviewdoc.cpp:245
    #13 0x81df4a in CMainFrame::OnOpen D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:669
    #14 0x5935fbcd  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x101efbcd)
    #15 0x59407a4d  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10297a4d)
    #16 0x81a5ec in CMainFrame::OnCmdMsg D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:195
    #17 0x593fa978  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x1028a978)
    #18 0x59406d2c  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10296d2c)
    #19 0x81a9e2 in CMainFrame::OnCommand D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:216
    #20 0x593f9f48  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289f48)
    #21 0x593f9ecc  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10289ecc)
    #22 0x8213b5 in CMainFrame::WindowProc D:\Projects\TheAssemblyArmada\Thyme\src\tools\w3dviewer\mainfrm.cpp:229
    #23 0x593f8183  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10288183)
    #24 0x593f83d3  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x102883d3)
    #25 0x592d7653  (C:\WINDOWS\SYSTEM32\mfc140.dll+0x10167653)
    #26 0x763716ea in AddClipboardFormatListener+0x4a (C:\WINDOWS\System32\USER32.dll+0x69e416ea)
    #27 0x76367b39 in GetClassLongW+0x7a9 (C:\WINDOWS\System32\USER32.dll+0x69e37b39)

SUMMARY: AddressSanitizer: heap-use-after-free D:\Projects\TheAssemblyArmada\Thyme\src\w3d\renderer\assetmgr.cpp:404 in W3DAssetManager::Release_All_Textures
Shadow bytes around the buggy address:
  0x32f8cea0: fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa
  0x32f8ceb0: fa fa fa fa fa fa fd fd fd fd fd fd fd fd fd fd
  0x32f8cec0: fd fd fd fd fa fa fa fa fa fa fa fa fd fd fd fd
  0x32f8ced0: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
  0x32f8cee0: fa fa fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x32f8cef0: fa fa fa fa fa fa fa fa[fd]fd fd fd fd fd fd fd
  0x32f8cf00: fd fd fd fd fd fd fa fa fa fa fa fa fa fa fd fd
  0x32f8cf10: fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa
  0x32f8cf20: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x32f8cf30: fd fd fa fa fa fa fa fa fa fa fd fd fd fd fd fd
  0x32f8cf40: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
Address Sanitizer Error: Use of deallocated memory
```

Reply:
Release_Ref call is there in stock so something else must be going wrong somewhere else that is causing it to fail.

Reply:
Fixed